lib/core/src/connection_pool.cpp

```diff
@@ -237,11 +283,45 @@ namespace irods
         }
 
         try {
-            // NOLINTNEXTLINE(bugprone-unused-raii)
-            query<RcComm>{ctx.conn.get(), "select ZONE_NAME where ZONE_TYPE = 'local'"};
-            if (std::time(nullptr) - ctx.creation_time > refresh_time_) {
-                return false;
+            // Check if the number of retrievals via get_connection() has been reached.
+            if (options_.number_of_retrievals_before_connection_refresh) {
+                if (ctx.retrieval_count >= *options_.number_of_retrievals_before_connection_refresh) {
+                    return false;
+                }
+            }
+
+            // Check age of connection.
+            if (options_.number_of_seconds_before_connection_refresh) {
+                const auto elapsed = std::chrono::steady_clock::now() - ctx.creation_time;
+
+                if (elapsed >= *options_.number_of_seconds_before_connection_refresh) {
+                    return false;
+                }
+            }
+
+            // Check if any resources have been modified.
+            if (options_.refresh_connections_when_resource_changes_detected) {
+                for (auto&& row : irods::query{ctx.conn.get(), "select max(RESC_MODIFY_TIME)"}) {
+                    try {
+                        if (const auto mtime = std::stoull(row[0]); mtime > ctx.latest_resc_mtime) {
+                            ctx.latest_resc_mtime = mtime;
+                            return false;
+                        }
+                    }
+                    catch (...) {
+                        // In theory, we should never reach this block. However, if we do, just
+                        // go ahead and refresh the connection. At the very least, the agent will
+                        // reflect the latest state in the catalog for a small penalty of TCP socket
+                        // reconstruction.
```

Comment:
let's log something horrible as well, so we can see if it ever happens.

Reply:
Can't log in this class. It is a client-side library.

Reply:
totally knew that.    of course.    carry on.